File: Java/src/test/java/workshops/testingil/dirtytests/DirtytestsApplicationTests.java
package workshops.testingil.dirtytests;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;

import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.web.client.HttpStatusCodeException;
import org.springframework.web.client.RestTemplate;

import com.fasterxml.jackson.databind.ObjectMapper;

import testingil.webinar.cleantests.CalculatorParams;
import testingil.webinar.cleantests.Ops;

@SpringBootTest
class DirtytestsApplicationTests {

	// Check that we can add two numbers
	// and return the right result as a string
	@Test
	void test_add() throws Exception {
		CalculatorParams calcParams = new CalculatorParams();
		calcParams.setFirst(3);
		calcParams.setSecond(4);
		calcParams.setfirstOp(OperationType.Plus);

		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		ObjectMapper mapper = new ObjectMapper();
		String json = mapper.writeValueAsString(calcParams);

		HttpEntity<String> request = new HttpEntity<String>(json, headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/calculate", request, String.class);

		assertThat(result, is("7"));
	}

	// Check that we can add two numbers
	// and return the right result as a string
	@Test
	void test_add_minus() throws Exception {
		CalculatorParams calcParams = new CalculatorParams();

		calcParams.setFirst(-5);
		calcParams.setSecond(-4);
		calcParams.setfirstOp(OperationType.Plus);

		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		ObjectMapper mapper = new ObjectMapper();
		String json = mapper.writeValueAsString(calcParams);

		HttpEntity<String> request = new HttpEntity<String>(json, headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/calculate", request, String.class);

		assertThat(result, is("-9"));

	}

	// subtract numbers
	@Test
	void test_minus() throws Exception {
		CalculatorParams calcParams = new CalculatorParams();

		calcParams.setFirst(20);
		calcParams.setSecond(4);
		calcParams.setfirstOp(OperationType.Minus);

		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		ObjectMapper mapper = new ObjectMapper();
		String json = mapper.writeValueAsString(calcParams);

		HttpEntity<String> request = new HttpEntity<String>(json, headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/calculate", request, String.class);

		assertThat(result, is("16"));
	}

	@Test
	// multiply two numbers
	void test_mul() throws Exception {
		CalculatorParams calcParams = new CalculatorParams();

		calcParams.setFirst(3);
		calcParams.setSecond(6);
		calcParams.setfirstOp(OperationType.Multiply);

		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		ObjectMapper mapper = new ObjectMapper();
		String json = mapper.writeValueAsString(calcParams);

		HttpEntity<String> request = new HttpEntity<String>(json, headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/calculate", request, String.class);

		assertThat(result, is("18"));
	}

	@Test
	void test_reset() throws Exception {
		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		HttpEntity<String> request = new HttpEntity<String>("", headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/reset", request, String.class);

		assertThat(result, is("0"));
	}

	@Test
	// Divide numbers
	void test_div() throws Exception {
		CalculatorParams calcParams = new CalculatorParams();

		calcParams.setFirst(6);
		calcParams.setSecond(3);
		calcParams.setfirstOp(OperationType.Div);

		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		ObjectMapper mapper = new ObjectMapper();
		String json = mapper.writeValueAsString(calcParams);

		HttpEntity<String> request = new HttpEntity<String>(json, headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/calculate", request, String.class);

		assertThat(result, is("2"));
	}

	@Test
	// Divide numbers
	void test_div_errors() throws Exception {
		CalculatorParams calcParams = new CalculatorParams();

		calcParams.setFirst(6);
		calcParams.setSecond(0);
		calcParams.setfirstOp(OperationType.Div);

		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		ObjectMapper mapper = new ObjectMapper();
		String json = mapper.writeValueAsString(calcParams);

		HttpEntity<String> request = new HttpEntity<String>(json, headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/calculate", request, String.class);

		assertThat(result, is("E"));
	}

	@Test
	// multiple numbers
	void test_mul_numbers() throws Exception {
		CalculatorParams calcParams = new CalculatorParams();

		calcParams.setFirst(3);
		calcParams.setSecond(4);
		calcParams.setSecond(7);
		calcParams.setfirstOp(OperationType.Plus);
		calcParams.setSecondOp(OperationType.Minus);

		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		ObjectMapper mapper = new ObjectMapper();
		String json = mapper.writeValueAsString(calcParams);

		HttpEntity<String> request = new HttpEntity<String>(json, headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/calculate", request, String.class);

		assertThat(result, is("0"));
	}

	@Test
	// multiple numbers
	void test_mul_plus() throws Exception {
		CalculatorParams calcParams = new CalculatorParams();

		calcParams.setFirst(3);
		calcParams.setSecond(4);
		calcParams.setSecond(2);
		calcParams.setfirstOp(OperationType.Plus);
		calcParams.setSecondOp(OperationType.Div);

		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		ObjectMapper mapper = new ObjectMapper();
		String json = mapper.writeValueAsString(calcParams);

		HttpEntity<String> request = new HttpEntity<String>(json, headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/calculate", request, String.class);

		assertThat(result, is("5"));
	}

	@Test
	// multiple numbers
	void test_mul_minus() throws Exception {
		CalculatorParams calcParams = new CalculatorParams();

		calcParams.setFirst(3);
		calcParams.setSecond(4);
		calcParams.setSecond(2);
		calcParams.setfirstOp(OperationType.Multiply);
		calcParams.setSecondOp(OperationType.Minus);

		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		ObjectMapper mapper = new ObjectMapper();
		String json = mapper.writeValueAsString(calcParams);

		HttpEntity<String> request = new HttpEntity<String>(json, headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/calculate", request, String.class);

		assertThat(result, is("10"));
	}

	
	@Test
	void test_reset_2() throws Exception {
		HttpHeaders headers = new HttpHeaders();
		headers.setContentType(MediaType.APPLICATION_JSON);

		HttpEntity<String> request = new HttpEntity<String>("", headers);

		RestTemplate restTemplate = new RestTemplate();
		String result = restTemplate.postForObject("http://localhost:8888/root/reset", request, String.class);

		try {
			result =  restTemplate.postForObject("http://localhost:8888/root/reset", request, String.class);
		} catch (HttpStatusCodeException e) {
			assertThat(e.getRawStatusCode(), is(400));
		}
	}
}

File: C#/DirtyTests/Exercise/DirtytestsApplicationTests.cs
using DirtyTestsWorkshop;
using DirtyTestsWorkshop.Demo;
using DirtyTestsWorkshop.Exercise;
using Microsoft.AspNetCore.Mvc.Testing;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Newtonsoft.Json;
using System.Net.Http;
using System.Text;
using System.Threading.Tasks;
using System.Net;

namespace DirtyTests
{
    [TestClass]
    public class DirtytestsApplicationTests
    {
        // Check that we can add two numbers
        // and return the right result as a string
        [TestMethod]
        public async Task Test_Add()
        {
            CalculatorParams calcParams = new CalculatorParams();
            calcParams.SetFirst(3);
            calcParams.SetSecond(4);
            calcParams.SetFirstOp(OperationType.Plus);

            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            string json = JsonConvert.SerializeObject(calcParams);
            HttpContent content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/calculate", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
            string result = await response.Content.ReadAsStringAsync();
            Assert.AreEqual("7", result);
        }

        // Check that we can add two numbers
        // and return the right result as a string
        [TestMethod]
        public async Task Test_Add_Minus()
        {
            CalculatorParams calcParams = new CalculatorParams();
            calcParams.SetFirst(-5);
            calcParams.SetSecond(-4);
            calcParams.SetFirstOp(OperationType.Plus);

            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            string json = JsonConvert.SerializeObject(calcParams);
            HttpContent content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/calculate", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
            string result = await response.Content.ReadAsStringAsync();

            Assert.AreEqual("-9", result);
        }
        // subtract numbers
        [TestMethod]
        public async Task Test_Minus()
        {
            CalculatorParams calcParams = new CalculatorParams();
            calcParams.SetFirst(20);
            calcParams.SetSecond(4);
            calcParams.SetFirstOp(OperationType.Minus);

            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            string json = JsonConvert.SerializeObject(calcParams);
            HttpContent content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/calculate", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
            string result = await response.Content.ReadAsStringAsync();

            Assert.AreEqual("16", result);
        }

        [TestMethod]
        // multiply two numbers
        public async Task Test_Mul()
        {
            CalculatorParams calcParams = new CalculatorParams();
            calcParams.SetFirst(3);
            calcParams.SetSecond(6);
            calcParams.SetFirstOp(OperationType.Multiply);

            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            string json = JsonConvert.SerializeObject(calcParams);
            HttpContent content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/calculate", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
            string result = await response.Content.ReadAsStringAsync();

            Assert.AreEqual("18", result);
        }

        [TestMethod]
        public async Task Test_Reset()
        {
            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            HttpContent content = new StringContent("", Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/reset", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
            string result = await response.Content.ReadAsStringAsync();

            Assert.AreEqual("0", result);
        }

        [TestMethod]
        // Divide numbers
        public async Task Test_Div()
        {
            CalculatorParams calcParams = new CalculatorParams();
            calcParams.SetFirst(6);
            calcParams.SetSecond(3);
            calcParams.SetFirstOp(OperationType.Div);

            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            string json = JsonConvert.SerializeObject(calcParams);
            HttpContent content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/calculate", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
            string result = await response.Content.ReadAsStringAsync();

            Assert.AreEqual("2", result);

        }


        [TestMethod]
        // Divide numbers
        public async Task Test_Div_errors()
        {
            CalculatorParams calcParams = new CalculatorParams();
            calcParams.SetFirst(6);
            calcParams.SetSecond(0);
            calcParams.SetFirstOp(OperationType.Div);

            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            string json = JsonConvert.SerializeObject(calcParams);
            HttpContent content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/calculate", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
            string result = await response.Content.ReadAsStringAsync();

            Assert.AreEqual("E", result);

        }

        [TestMethod]
        // multiple numbers
        public async Task Test_Mul_Numbers()
        {
            CalculatorParams calcParams = new CalculatorParams();
            calcParams.SetFirst(3);
            calcParams.SetSecond(4);
            calcParams.SetThird(7);
            calcParams.SetFirstOp(OperationType.Plus);
            calcParams.SetSecondOp(OperationType.Minus);

            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            string json = JsonConvert.SerializeObject(calcParams);
            HttpContent content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/calculate", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
            string result = await response.Content.ReadAsStringAsync();

            Assert.AreEqual("0", result);
        }
        [TestMethod]
        // multiple numbers
        public async Task Test_Mul_Plus()
        {
            CalculatorParams calcParams = new CalculatorParams();
            calcParams.SetFirst(3);
            calcParams.SetSecond(4);
            calcParams.SetThird(2);
            calcParams.SetFirstOp(OperationType.Plus);
            calcParams.SetSecondOp(OperationType.Div);

            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            string json = JsonConvert.SerializeObject(calcParams);
            HttpContent content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/calculate", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
            string result = await response.Content.ReadAsStringAsync();

            Assert.AreEqual("5", result);
        }

        [TestMethod]
        // multiple numbers
        public async Task Test_Mul_Minus()
        {
            CalculatorParams calcParams = new CalculatorParams();

            calcParams.SetFirst(3);
            calcParams.SetSecond(4);
            calcParams.SetThird(2);
            calcParams.SetFirstOp(OperationType.Multiply);
            calcParams.SetSecondOp(OperationType.Minus);

            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            string json = JsonConvert.SerializeObject(calcParams);
            HttpContent content = new StringContent(json, Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/calculate", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
            string result = await response.Content.ReadAsStringAsync();

            Assert.AreEqual("10", result);
        }


        [TestMethod]
        public async Task Test_Reset_2()
        {
            WebApplicationFactory<Startup> appFactory = new WebApplicationFactory<Startup>();
            HttpClient _client = appFactory.CreateClient();

            HttpContent content = new StringContent("", Encoding.UTF8, "application/json");

            var response = await _client.PostAsync("http://localhost:8888/root/reset", content);
            Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);

            response = await _client.PostAsync("http://localhost:8888/root/reset", content);
            Assert.AreEqual(HttpStatusCode.BadRequest, response.StatusCode);
        }

    }
}
